crates/ra_assists/src/add_missing_impl_members.rs

```diff
@@ -0,0 +1,279 @@
+use crate::{Assist, AssistId, AssistCtx};
+
+use hir::Resolver;
+use hir::db::HirDatabase;
+use ra_syntax::{SmolStr, SyntaxKind, TextRange, TextUnit, TreeArc};
+use ra_syntax::ast::{self, AstNode, FnDef, ImplItem, ImplItemKind, NameOwner};
+use ra_db::FilePosition;
+use ra_fmt::{leading_indent, reindent};
+
+use itertools::Itertools;
+
+/// Given an `ast::ImplBlock`, resolves the target trait (the one being
+/// implemented) to a `ast::TraitDef`.
+pub(crate) fn resolve_target_trait_def(
+    db: &impl HirDatabase,
+    resolver: &Resolver,
+    impl_block: &ast::ImplBlock,
+) -> Option<TreeArc<ast::TraitDef>> {
+    let ast_path = impl_block.target_trait().map(AstNode::syntax).and_then(ast::PathType::cast)?;
+    let hir_path = ast_path.path().and_then(hir::Path::from_ast)?;
+
+    match resolver.resolve_path(db, &hir_path).take_types() {
+        Some(hir::Resolution::Def(hir::ModuleDef::Trait(def))) => Some(def.source(db).1),
+        _ => None,
+    }
+}
+
+pub(crate) fn build_func_body(def: &ast::FnDef) -> String {
+    let mut buf = String::new();
+
+    for child in def.syntax().children() {
+        if child.kind() == SyntaxKind::SEMI {
+            buf.push_str(" { unimplemented!() }")
+        } else {
+            child.text().push_to(&mut buf);
+        }
+    }
+
+    buf.trim_end().to_string()
+}
+
+pub(crate) fn add_missing_impl_members(mut ctx: AssistCtx<impl HirDatabase>) -> Option<Assist> {
+    let node = ctx.covering_node();
+    let impl_node = node.ancestors().find_map(ast::ImplBlock::cast)?;
+    let impl_item_list = impl_node.item_list()?;
+    // Don't offer the assist when cursor is at the end, outside the block itself.
+    let cursor_range = TextRange::from_to(node.range().end(), node.range().end());
+    if node.range().end() == impl_node.syntax().range().end() {
+        return None;
+    }
+
+    let trait_def = {
+        let position = FilePosition { file_id: ctx.frange.file_id, offset: cursor_range.end() };
+        let resolver = hir::source_binder::resolver_for_position(ctx.db, position);
+
+        resolve_target_trait_def(ctx.db, &resolver, impl_node)?
+    };
+
+    let missing_fns: Vec<_> = {
+        let fn_def_opt = |kind| if let ImplItemKind::FnDef(def) = kind { Some(def) } else { None };
+        let def_name = |def| -> Option<&SmolStr> { FnDef::name(def).map(ast::Name::text) };
+
+        let trait_items =
+            trait_def.syntax().descendants().find_map(ast::ItemList::cast)?.impl_items();
+        let impl_items = impl_item_list.impl_items();
+
+        let trait_fns = trait_items.map(ImplItem::kind).filter_map(fn_def_opt).collect::<Vec<_>>();
+        let impl_fns = impl_items.map(ImplItem::kind).filter_map(fn_def_opt).collect::<Vec<_>>();
+
+        trait_fns
+            .into_iter()
+            .filter(|t| def_name(t).is_some())
+            .filter(|t| impl_fns.iter().all(|i| def_name(i) != def_name(t)))
```

Comment:
Originally I went for the [hash set difference](https://github.com/rust-analyzer/rust-analyzer/pull/947/commits/149b2f540abfcb6de70dbe194dbf77fb3bb545b4) between sets of function names but figured out this would be faster and with less allocations in the regular case of < 10 functions per trait/impl block

Reply:
Yeah, this looks right. In general, these refactoring bits are notoriously hard to estimate in terms of impact on the performance. They usually contain a ton of code for handling all kinds of special cases, so they feel like they should be performance-sensitive just from the volume of the code. However, when you factor in the fact that this is computed only for a single position, when the cursor is inside the impl, and that the amount of data this needs to process is tiny, it becomes clear that performance is not a concern.